### PR TITLE
Reduce DataTypeImpl binary size

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Reduce the number of virtual methods in DataTypeImpl to reduce binary size.
Refactor some helpers to reduce the amount of templatized code.

**Motivation and Context**
Counter binary size increases from new types. 

>./build.bat --config MinSizeRel --build_dir "build\Android\arm64.static\1711.NoOps" --android --android_sdk_path D:\Android --android_ndk_path D:\Android\ndk\21.1.6352462\ --android_abi arm64-v8a --android_api 29 --cmake_generator Ninja  --disable_ml_ops --disable_exceptions --build_shared_lib --skip_tests --minimal_build --include_ops_by_config .\no_ops.config --cmake_extra_defines onnxruntime_DISABLE_SPARSE_TENSORS=ON onnxruntime_ENABLE_ORT_FORMAT_RUNTIME_GRAPH_OPTIMIZATION=OFF 

master: 1,212,232	
with changes:1,183,560

reduction: 28KB

### CODE REVIEW

File: include/onnxruntime/core/framework/data_types.h [hunk 1]
```diff
@@ -76,6 +76,22 @@ using CreateFunc = void* (*)();
  *
  */
 class DataTypeImpl {
+ public:
+  enum GeneralType {
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/data_types.h:80]: these look like flag values - are they intended to support being combined? e.g., for an optional tensor or optional tensor sequence

any reason for skipping the value `1 << 0`?

File: include/onnxruntime/core/framework/data_types.h [hunk 2]
```diff
@@ -936,7 +918,7 @@ class PrimitiveDataTypeBase : public DataTypeImpl {
   }
 
  protected:
-  PrimitiveDataTypeBase() = default;
+  PrimitiveDataTypeBase(size_t size) : DataTypeImpl{GeneralType::kNonTensor, size}, data_type_{-1} {}
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/data_types.h:921]: what does the data_type_ value of -1 mean?

File: include/onnxruntime/core/framework/data_types.h (hunk 1 above)
Comment (skottmckay) [reply include/onnxruntime/core/framework/data_types.h:80]: Good point. They don't need to be combined.

File: include/onnxruntime/core/framework/data_types.h (hunk 2 above)
Comment (skottmckay) [reply include/onnxruntime/core/framework/data_types.h]: Will update and remove SetDataType.

**Review by edgchen1 — DISMISSED** [review]

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**